File: sdks/typescript/core/src/protocol.ts
/**
 * Tesseron wire-protocol version sent in {@link HelloParams.protocolVersion}.
 * Major-version mismatches are hard-rejected by the gateway; minor-version
 * mismatches log a warning.
 *
 * 1.2.0 introduces the host-mint claim flow (tesseron#60): the
 * `tesseron-bind.<code>` WebSocket subprotocol element, the
 * `tesseron/bind` JSON-RPC method (UDS), `hostMintedClaim` in the
 * instance manifest, and `tesseron/claimed.agentCapabilities`.
 */
export const PROTOCOL_VERSION = '1.2.0' as const;
/** JSON-RPC version used for every message. */
export const JSONRPC_VERSION = '2.0' as const;

export type JsonRpcId = string | number | null;

export interface JsonRpcRequest<P = unknown> {
  jsonrpc: typeof JSONRPC_VERSION;
  id: JsonRpcId;
  method: string;
  params?: P;
}

export interface JsonRpcNotification<P = unknown> {
  jsonrpc: typeof JSONRPC_VERSION;
  method: string;
  params?: P;
}

export interface JsonRpcSuccessResponse<R = unknown> {
  jsonrpc: typeof JSONRPC_VERSION;
  id: JsonRpcId;
  result: R;
}

export interface JsonRpcErrorResponse {
  jsonrpc: typeof JSONRPC_VERSION;
  id: JsonRpcId;
  error: JsonRpcErrorPayload;
}

export type JsonRpcResponse<R = unknown> = JsonRpcSuccessResponse<R> | JsonRpcErrorResponse;

export interface JsonRpcErrorPayload {
  code: number;
  message: string;
  data?: unknown;
}

export type JsonRpcMessage<P = unknown, R = unknown> =
  | JsonRpcRequest<P>
  | JsonRpcNotification<P>
  | JsonRpcResponse<R>;

/**
 * Numeric codes for every error the SDK raises. Codes in `-32xxx` follow the
 * JSON-RPC reserved range; Tesseron-specific codes start at `-32000`. Use
 * {@link TesseronError.code} to branch on these.
 */
export const TesseronErrorCode = {
  ParseError: -32700,
  InvalidRequest: -32600,
  MethodNotFound: -32601,
  InvalidParams: -32602,
  InternalError: -32603,
  ProtocolMismatch: -32000,
  Cancelled: -32001,
  Timeout: -32002,
  ActionNotFound: -32003,
  InputValidation: -32004,
  HandlerError: -32005,
  SamplingNotAvailable: -32006,
  ElicitationNotAvailable: -32007,
  SamplingDepthExceeded: -32008,
  Unauthorized: -32009,
  TransportClosed: -32010,
  ResumeFailed: -32011,
} as const;

export type TesseronErrorCodeValue = (typeof TesseronErrorCode)[keyof typeof TesseronErrorCode];

/** Feature flags exchanged during the `tesseron/hello` handshake. */
export interface TesseronCapabilities {
  /** `true` if the peer can send/receive progress and log notifications during an invocation. */
  streaming: boolean;
  /** `true` if the peer honours `resources/subscribe`. */
  subscriptions: boolean;
  /** `true` if the peer can service `sampling/request`. On the gateway side, mirrors the MCP client's `sampling` capability. */
  sampling: boolean;
  /** `true` if the peer can service `elicitation/request`. On the gateway side, mirrors the MCP client's `elicitation` capability. */
  elicitation: boolean;
}

/** App identity as carried on the wire; differs from {@link AppConfig} in that `origin` is required. */
export interface AppMetadata {
  id: string;
  name: string;
  description?: string;
  origin: string;
  iconUrl?: string;
  version?: string;
}

/** MCP-aligned tool hints surfaced in the action manifest. */
export interface ActionAnnotations {
  /** `true` if the action has no side effects (agents may skip confirmation UIs). */
  readOnly?: boolean;
  /** `true` if the action can destroy or mutate state in a non-idempotent way. */
  destructive?: boolean;
  /** `true` if the client should confirm with the user before invoking. */
  requiresConfirmation?: boolean;
}

export interface ActionManifestEntry {
  name: string;
  description: string;
  inputSchema: unknown;
  outputSchema?: unknown;
  annotations?: ActionAnnotations;
  timeoutMs?: number;
}

export interface ResourceManifestEntry {
  name: string;
  description: string;
  outputSchema?: unknown;
  subscribable: boolean;
}

/** Parameters of the `tesseron/hello` request sent by the SDK on connect. */
export interface HelloParams {
  protocolVersion: string;
  app: AppMetadata;
  actions: ActionManifestEntry[];
  resources: ResourceManifestEntry[];
  capabilities: TesseronCapabilities;
}

/**
 * Result of the `tesseron/hello` handshake (and `tesseron/resume`). Carries the
 * session id, the capabilities the MCP client will honour, the connected
 * agent's identity (filled once a bridge attaches), the `claimCode` the user
 * pastes into their MCP client to link this session, and an opaque
 * `resumeToken` the caller can stash wherever fits their app to rejoin this
 * session after a transport drop.
 */
export interface WelcomeResult {
  sessionId: string;
  protocolVersion: string;
  capabilities: TesseronCapabilities;
  agent: AgentIdentity;
  /**
   * 6-character pairing code the user enters in their MCP client. Present on
   * every `tesseron/hello` response. Always **absent** on a successful
   * `tesseron/resume` response: the gateway only permits resume for
   * already-claimed sessions, so re-issuing a pairing code would be a no-op
   * at best and a UI confusion at worst.
   */
  claimCode?: string;
  /**
   * Server-issued token that, together with {@link WelcomeResult.sessionId},
   * lets a reconnecting SDK rejoin this session via `tesseron/resume` after the
   * transport drops. Rotated on every successful resume (one-shot) — callers
   * that persist it MUST overwrite with the freshest value from each handshake.
   *
   * Optional on the wire for backwards-compatibility with gateways that pre-date
   * the resume extension; new gateways always populate it.
   */
  resumeToken?: string;
}

/** Identity advertised by the MCP client that claimed this session. */
export interface AgentIdentity {
  id: string;
  name: string;
}

/**
 * Parameters of the `tesseron/resume` request sent by the SDK when it already
 * has a `sessionId` + `resumeToken` pair from a prior handshake and wants to
 * rejoin an existing (possibly-claimed) session instead of opening a fresh one.
 *
 * Carries the same `app` / `actions` / `resources` / `capabilities` as a
 * regular `tesseron/hello` because a fresh app build may have added, removed,
 * or changed them since the previous connect. The gateway updates the session's
 * registered manifest from these fields on a successful resume.
 */
export interface ResumeParams extends HelloParams {
  /** Opaque session identifier returned in the prior {@link WelcomeResult}. */
  sessionId: string;
  /**
   * Opaque token returned in the prior {@link WelcomeResult}. Gateway compares
   * against the stored token in constant time and rotates it on success; the
   * freshly-returned {@link WelcomeResult.resumeToken} is the one to persist.
   */
  resumeToken: string;
}

/**
 * Result of the `tesseron/resume` request. Same shape as {@link WelcomeResult}
 * — `sessionId` matches the param, `resumeToken` is rotated, and `claimCode`
 * is always absent (the gateway only permits resume for already-claimed
 * sessions, so there is no pairing code to re-issue).
 */
export type ResumeResult = WelcomeResult;

export interface ActionInvokeParams {
  name: string;
  input: unknown;
  invocationId: string;
  client?: { route?: string };
}

export interface ActionProgressParams {
  invocationId: string;
  message?: string;
  percent?: number;
  data?: unknown;
}

export interface ActionResultPayload {
  invocationId: string;
  output: unknown;
}

export interface ActionCancelParams {
  invocationId: string;
}

export interface ResourceReadParams {
  name: string;
}

export interface ResourceReadResult {
  value: unknown;
}

export interface ResourceSubscribeParams {
  name: string;
  subscriptionId: string;
}

export interface ResourceUnsubscribeParams {
  subscriptionId: string;
}

export interface ResourceUpdatedParams {
  subscriptionId: string;
  value: unknown;
}

export interface SamplingRequestParams {
  invocationId: string;
  prompt: string;
  schema?: unknown;
  maxTokens?: number;
}

export interface SamplingResult {
  content: unknown;
}

export interface ElicitationRequestParams {
  invocationId: string;
  question: string;
  schema: unknown;
}

export interface ElicitationResult {
  /**
   * Mirrors the MCP elicit `action`: `accept` carries `value`; `decline` and
   * `cancel` do not. SDK callers use `ctx.confirm` for yes/no (action-only)
   * or `ctx.elicit` for structured content (action + value).
   */
  action: 'accept' | 'decline' | 'cancel';
  value?: unknown;
}

export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

export interface LogParams {
  invocationId?: string;
  level: LogLevel;
  message: string;
  meta?: Record<string, unknown>;
}

/**
 * Parameters of the `tesseron/bind` request a v1.2 gateway sends as the
 * very first frame after a UDS connect, when dialing a host-minted
 * manifest. Replaces the WebSocket `tesseron-bind.<code>` subprotocol
 * element on transports without a subprotocol concept of their own.
 *
 * The host validates `code` against its in-memory `hostMintedClaim.code`
 * in constant time. On success: replies `{ ok: true }`, marks the bind
 * authoritative, and proceeds with the v3 hello-replay flow. On
 * mismatch: replies with `Unauthorized` and closes the channel.
 */
export interface BindParams {
  code: string;
}

/** Response to `tesseron/bind`. Only `{ ok: true }` is meaningful. */
export interface BindResult {
  ok: true;
}

export interface TesseronMethods {
  'tesseron/hello': { params: HelloParams; result: WelcomeResult };
  'tesseron/resume': { params: ResumeParams; result: ResumeResult };
  'tesseron/bind': { params: BindParams; result: BindResult };
  'sampling/request': { params: SamplingRequestParams; result: SamplingResult };
  'elicitation/request': { params: ElicitationRequestParams; result: ElicitationResult };
  'actions/invoke': { params: ActionInvokeParams; result: ActionResultPayload };
  'resources/read': { params: ResourceReadParams; result: ResourceReadResult };
  'resources/subscribe': { params: ResourceSubscribeParams; result: undefined };
  'resources/unsubscribe': { params: ResourceUnsubscribeParams; result: undefined };
}

/**
 * Parameters of the `tesseron/claimed` notification sent by the gateway to the
 * SDK when an agent calls `tesseron__claim_session` on a session that was
 * previously in the awaiting-claim state. After this notification fires, the
 * session's `claimCode` is consumed and no longer claimable; consumers
 * displaying the code should clear it.
 */
export interface ClaimedParams {
  /** Identity of the agent that just claimed this session. */
  agent: AgentIdentity;
  /** Unix epoch milliseconds at which the gateway processed the claim. */
  claimedAt: number;
  /**
   * Negotiated capability bits the gateway will honour for this session
   * (sampling / elicitation depend on whether the attached MCP client
   * advertised them; streaming / subscriptions are always `true`). When
   * present, the SDK overwrites `WelcomeResult.capabilities` so action
   * handlers gating on `ctx.agentCapabilities.sampling` see authoritative
   * values rather than the SDK's own pre-claim defaults.
   *
   * Optional for back-compat with v1.1 gateways that didn't carry the
   * field; v1.1 SDKs that don't see it keep using the welcome-time
   * capabilities, which were authoritative in v1.1 because the gateway
   * minted the welcome itself. Required-in-spirit for v1.2 hosts that
   * synthesize the welcome — without it, `ctx.agentCapabilities`
   * reports the SDK's *own* capabilities, not the gateway's. See
   * tesseron#60.
   */
  agentCapabilities?: TesseronCapabilities;
}

export interface TesseronNotifications {
  'actions/progress': ActionProgressParams;
  'actions/cancel': ActionCancelParams;
  'actions/list_changed': { actions: ActionManifestEntry[] };
  'resources/list_changed': { resources: ResourceManifestEntry[] };
  'resources/updated': ResourceUpdatedParams;
  'tesseron/claimed': ClaimedParams;
  log: LogParams;
}

File: sdks/python/src/tesseron/protocol.py
"""The Tesseron 1.2 wire format: method names, envelope shapes, and the manifest models."""

from __future__ import annotations

from enum import StrEnum
from typing import Final

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from .json_types import JsonObject, JsonValue

__all__ = [
    "GATEWAY_SUBPROTOCOL",
    "JSONRPC_VERSION",
    "PROTOCOL_VERSION",
    "ActionDescriptor",
    "AgentIdentity",
    "ApplicationDescriptor",
    "Capabilities",
    "ClaimedParams",
    "HelloParams",
    "LogLevel",
    "Methods",
    "ResourceDescriptor",
    "ResumeParams",
    "WelcomeResult",
    "is_valid_application_id",
    "shares_major_version",
]

PROTOCOL_VERSION: Final = "1.2.0"
"""The protocol version this package speaks, as it appears in every handshake."""

JSONRPC_VERSION: Final = "2.0"
"""The JSON-RPC version every envelope carries."""

GATEWAY_SUBPROTOCOL: Final = "tesseron-gateway"
"""The WebSocket subprotocol the gateway sends on its upgrade request.

An upgrade without it is not a gateway dial, and the host refuses it: the endpoint exists
for the gateway, not for arbitrary local clients.
"""


class Methods:
    """The JSON-RPC method names this package sends or answers."""

    HELLO: Final = "tesseron/hello"
    RESUME: Final = "tesseron/resume"
    CLAIMED: Final = "tesseron/claimed"
    INVOKE: Final = "actions/invoke"
    CANCEL: Final = "actions/cancel"
    PROGRESS: Final = "actions/progress"
    READ: Final = "resources/read"
    SUBSCRIBE: Final = "resources/subscribe"
    UNSUBSCRIBE: Final = "resources/unsubscribe"
    UPDATED: Final = "resources/updated"
    SAMPLE: Final = "sampling/request"
    ELICIT: Final = "elicitation/request"
    LOG: Final = "log"


class LogLevel(StrEnum):
    """Severity of a ``log`` notification, matching the MCP levels the gateway forwards to."""

    DEBUG = "debug"
    INFO = "info"
    WARN = "warn"
    ERROR = "error"


class WireModel(BaseModel):
    """A model whose Python field names are snake_case and whose JSON keys are camelCase."""

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True, extra="ignore")


class Capabilities(WireModel):
    """What the application itself can do, sent in the handshake.

    This is the application half of the negotiation. The welcome comes back with the
    intersection of these flags and the agent's own.
    """

    streaming: bool = False
    subscriptions: bool = False
    sampling: bool = False
    elicitation: bool = False

    @classmethod
    def implemented(cls) -> Capabilities:
        """The set this package can honestly declare: everything protocol 1.2.0 defines."""
        return cls(streaming=True, subscriptions=True, sampling=True, elicitation=True)

    @classmethod
    def none(cls) -> Capabilities:
        """Nothing negotiated: the answer a handler gets before a welcome arrives."""
        return cls()

    def to_wire(self) -> JsonObject:
        """The four negotiation flags, all of them always present."""
        return {
            "streaming": self.streaming,
            "subscriptions": self.subscriptions,
            "sampling": self.sampling,
            "elicitation": self.elicitation,
        }


class ApplicationDescriptor(WireModel):
    """Who the application is, as the gateway and the agent see it."""

    id: str
    name: str
    description: str | None = None
    origin: str = "unknown"
    version: str | None = None
    icon_url: str | None = None

    def to_wire(self) -> JsonObject:
        """The descriptor, with the optional members the application did not set left out."""
        payload: JsonObject = {"id": self.id, "name": self.name, "origin": self.origin}
        if self.description is not None:
            payload["description"] = self.description
        if self.version is not None:
            payload["version"] = self.version
        if self.icon_url is not None:
            payload["iconUrl"] = self.icon_url
        return payload


class ActionDescriptor(WireModel):
    """One action as it appears in the handshake manifest."""

    name: str
    description: str = ""
    input_schema: JsonValue = Field(default_factory=dict)
    output_schema: JsonValue = None
    timeout_ms: int | None = None

    def to_wire(self) -> JsonObject:
        """The descriptor.

        ``description`` and ``inputSchema`` are always sent, even when they are empty,
        because the gateway projects both into the MCP tool definition.
        """
        payload: JsonObject = {
            "name": self.name,
            "description": self.description,
            "inputSchema": {} if self.input_schema is None else self.input_schema,
        }
        if self.output_schema is not None:
            payload["outputSchema"] = self.output_schema
        if self.timeout_ms is not None:
            payload["timeoutMs"] = self.timeout_ms
        return payload


class ResourceDescriptor(WireModel):
    """One resource as it appears in the handshake manifest."""

    name: str
    description: str = ""
    subscribable: bool = False

    def to_wire(self) -> JsonObject:
        """The descriptor. Every member is always sent."""
        return {
            "name": self.name,
            "description": self.description,
            "subscribable": self.subscribable,
        }


class AgentIdentity(WireModel):
    """Identity of the agent on the other end of the session."""

    id: str
    name: str

    @classmethod
    def pending(cls) -> AgentIdentity:
        """The placeholder identity a session carries until a claim happens."""
        return cls(id="pending", name="Awaiting agent")


class WelcomeResult(WireModel):
    """The result of ``tesseron/hello`` and of ``tesseron/resume``.

    There is no ``tesseron/welcome`` method on the wire; "welcome" is the name of this
    result shape. A gateway that omits ``capabilities`` has negotiated nothing, so the
    default is the empty set rather than what this host declared: assuming the agent can
    sample or elicit would make handlers wait on requests nobody answers.
    """

    session_id: str
    protocol_version: str
    capabilities: Capabilities = Field(default_factory=Capabilities.none)
    agent: AgentIdentity = Field(default_factory=AgentIdentity.pending)
    claim_code: str | None = None
    resume_token: str | None = None


class ClaimedParams(WireModel):
    """``tesseron/claimed`` parameters: the claim code has been redeemed."""

    agent: AgentIdentity
    claimed_at: int = 0
    agent_capabilities: JsonValue = None


class HelloParams(WireModel):
    """``tesseron/hello`` parameters: the whole application manifest."""

    protocol_version: str
    app: ApplicationDescriptor
    actions: list[ActionDescriptor]
    resources: list[ResourceDescriptor]
    capabilities: Capabilities

    def to_wire(self) -> JsonObject:
        """The manifest, with each descriptor serialised by its own rules."""
        return {
            "protocolVersion": self.protocol_version,
            "app": self.app.to_wire(),
            "actions": [action.to_wire() for action in self.actions],
            "resources": [resource.to_wire() for resource in self.resources],
            "capabilities": self.capabilities.to_wire(),
        }


class ResumeParams(HelloParams):
    """``tesseron/resume`` parameters: the manifest again, plus the credentials.

    The manifest repeats because a restarted application may have added, removed, or
    changed actions since the session was claimed; the gateway replaces its stored copy
    with whatever resume brings in.
    """

    session_id: str
    resume_token: str

    def to_wire(self) -> JsonObject:
        """The manifest plus the credentials this resume presents."""
        payload = super().to_wire()
        payload["sessionId"] = self.session_id
        payload["resumeToken"] = self.resume_token
        return payload


RESERVED_APPLICATION_IDS: Final = frozenset({"tesseron", "mcp", "system"})


def is_valid_application_id(application_id: str) -> bool:
    """Whether an application id is usable as an MCP tool prefix.

    The grammar is ``^[a-z][a-z0-9_]*$`` and three ids are reserved for the gateway's own
    tools.
    """
    if not application_id or application_id in RESERVED_APPLICATION_IDS:
        return False
    if not ("a" <= application_id[0] <= "z"):
        return False
    return all(
        ("a" <= character <= "z") or ("0" <= character <= "9") or character == "_"
        for character in application_id[1:]
    )


def shares_major_version(left: str, right: str) -> bool:
    """Whether two protocol versions agree on their major component.

    A missing major is treated as a mismatch rather than as a match, because guessing is
    how a 2.x gateway silently talks to a 1.x host.
    """
    left_major = left.split(".", 1)[0]
    right_major = right.split(".", 1)[0]
    return bool(left_major) and left_major == right_major
